### DISCUSSION TITLE
JOSM plugin wrapping valhalla routing engine

### DISCUSSION THREAD

**Discussion by tsmock** [valhalla/valhalla#4974] [General]
I've been working on a JOSM plugin that wraps some binaries I've built (via GH actions).

A few questions:
1. Do you mind if users report bugs against the routing engine here (example: "Valhalla has instructions to turn left and then do an immediate u-turn when a right-turn would be OK, there was no turn restrictions, here is the osm file that exhibited the behavior")
2. Is there a way to get costs for all segments in a pbf? isochrone is _kind of_ what I want, except I want the costs of the segments. This is _not_ something I need an answer to "soon", but I think it would be useful for mappers.
3. Is there anything you need me to do in the plugin (e.g. links, callouts, etc.)
4. From _your_ perspective, is there anything useful that a JOSM plugin could do for you?

Notes on the _current_ implementation (2024-11-14):
* The plugin calls the following CLI programs:
  1. `valhalla_build_timezones` (but only if not already built)
  2. `valhalla_build_admins`
  3. `valhalla_build_tiles`
  4. `valhalla_build_extract`
  5. `valhalla_service`

The last 4 steps are done every time the user changes the data.

JOSM ticket: https://josm.openstreetmap.de/ticket/23904

As an FYI, the plugin is not planned to be valhalla only.

Semi-related discussion: #2737

**Comment by kevinkreiser** [comment]
1. currently one can use valhalla via osm.org and we do, from time to time, get issues as a result of that. i would say we dont mind people reporting issues here so long as its nothing to do with jsom and pretty much everything to do with the router itself (id assume this is the case)
2. this is a great question. we do return costing information per edge in a route with the osrm format response. we dont have an endpoint to get just a list of edges and their costs given a costing and a bounding box. that would be pretty interesting for debugging purposes.
3. we dont require any attribution but anything that can help users to avoid making spurious issues would be appreciated. it would disappointing to get issues in which the primary complaint is that there is an alternate route based on local knowledge that is better than the one prescribed. often its quite difficult to get these routes without real time traffic data which doesnt exist in the open yet.
4. the router benefits primarily from more accurate data in osm. speed limit information, proper highway tagging and turn restrictions are probably the most impactful when it comes to the router. if the user is mapping in an area with a lot of imported ways that may have poorly chosen highway tags or if there is missing speedlimit info adding those would be great.

regarding your implementation, `valhalla_build_admins` relies on the full compliment of admin relations members being there to work. if they arent there or are only there in part, those particular admins wont be built and as a result you will get strange routing (wrong side of the road, wrong defaults for access restrictions by highway tag etc).

**Comment by tsmock** [comment]
1. I'd hope this is the case as well, but (as your comment on `valhalla_build_admins` shows), it might not always be the case if I made a poor assumption.
2. I was hoping you had something like that for debugging purposes.
3. This is totally fair. This is actually the reason why I was hoping there was a way to get costing information for each way (or edge), since it would help users know why a route isn't taken.
4. This is fair, and there are already styles for mapping speedlimits, although the poor highway tags can be a bit more difficult to pick up without traffic info. Which (as noted) there isn't a good open source of.

Anyway, thanks for the quick response (and the hint on `valhalla_build_admins`). I'll have to look into what it needs.
